File: tty.c
// qemus riscv virt machine places UART at this address
unsigned char *uart = (unsigned char *)0x10000000;

void putchar(char c) {
  *uart = c;
  return;
}

void print(const char *str) {
  while (*str != '\0') {
    putchar(*str);
    str++;
  }
  return;
}

void println(const char *str) {
  while (*str != '\0') {
    putchar(*str);
    str++;
  }
  putchar('\n');
  putchar('\r');
  return;
}

File: tty.zig
// qemu's virt machine type places uart at this adress.
const uart: *volatile u8 = @ptrFromInt(0x10000000);

pub fn print(str: []const u8) void {
    for (str) |c| {
        uart.* = c;
    }
}

pub fn println(str: []const u8) void {
    print(str);
    print("\n\r");
}
